lib/uri_parser.js

```diff
@@ -384,6 +404,71 @@ function parseQueryString(query, options) {
   return Object.keys(result).length ? result : null;
 }
 
+/**
+ * Checks a query string for invalid tls options according to the URI options spec.
+ *
+ * @param {string} queryString The query string to check
+ * @throws {MongoParseError}
+ */
+function checkTLSOptions(queryString) {
+  const queryStringKeys = Object.keys(queryString);
+  if (
+    queryStringKeys.indexOf('tlsInsecure') !== -1 &&
+    (queryStringKeys.indexOf('tlsAllowInvalidCertificates') !== -1 ||
+      queryStringKeys.indexOf('tlsAllowInvalidHostnames') !== -1)
+  ) {
+    throw new MongoParseError(
+      'The tlsInsecure option cannot be used with tlsAllowInvalidCertificates or tlsAllowInvalidHostnames.'
+    );
+  }
+
+  if (queryStringKeys.indexOf('tls') !== -1 && queryStringKeys.indexOf('ssl') !== -1) {
```

Comment:
nit: stick `queryStringKeys.indexOf('tls')` and `queryStringKeys.indexOf('ssl') !== -1` in variables and reuse them.